app/src/models/popup.ts

```diff
@@ -24,351 +24,530 @@ import { ValidNotificationPullRequestReview } from '../lib/valid-notification-pu
+  repository: Repository
+  onSubmitCommitMessage: (context: ICommitContext) => Promise<boolean>
+}
+
+interface IMultiCommitOperationPopup extends IBasePopup {
+  type: PopupType.MultiCommitOperation
+  repository: Repository
+}
+
+interface IWarnLocalChangesBeforeUndoPopup extends IBasePopup {
+  type: PopupType.WarnLocalChangesBeforeUndo
+  repository: Repository
+  commit: Commit
+  isWorkingDirectoryClean: boolean
+}
+
+interface IWarningBeforeResetPopup extends IBasePopup {
+  type: PopupType.WarningBeforeReset
+  repository: Repository
+  commit: Commit
+}
+
+interface IInvalidatedTokenPopup extends IBasePopup {
+  type: PopupType.InvalidatedToken
+  account: Account
+}
+
+interface IAddSSHHostPopup extends IBasePopup {
+  type: PopupType.AddSSHHost
+  host: string
+  ip: string
+  keyType: string
+  fingerprint: string
+  onSubmit: (addHost: boolean) => void
+}
+
+interface ISSHKeyPassphrasePopup extends IBasePopup {
+  type: PopupType.SSHKeyPassphrase
+  keyPath: string
+  onSubmit: (passphrase: string | undefined, storePassphrase: boolean) => void
+}
+
+interface ISSHUserPasswordPopup extends IBasePopup {
+  type: PopupType.SSHUserPassword
+  username: string
+  onSubmit: (password: string | undefined, storePassword: boolean) => void
+}
+
+interface IPullRequestChecksFailedPopup extends IBasePopup {
+  type: PopupType.PullRequestChecksFailed
+  repository: RepositoryWithGitHubRepository
+  pullRequest: PullRequest
+  shouldChangeRepository: boolean
+  commitMessage: string
+  commitSha: string
+  checks: ReadonlyArray<IRefCheck>
+}
+
+interface ICICheckRunRerunPopup extends IBasePopup {
+  type: PopupType.CICheckRunRerun
+  checkRuns: ReadonlyArray<IRefCheck>
+  repository: GitHubRepository
+  prRef: string
+  failedOnly: boolean
+}
+
+interface IWarnForcePushPopup extends IBasePopup {
+  type: PopupType.WarnForcePush
+  operation: string
+  onBegin: () => void
+}
+
+interface IDiscardChangesRetryPopup extends IBasePopup {
+  type: PopupType.DiscardChangesRetry
+  retryAction: RetryAction
+}
+
+interface IPullRequestReviewPopup extends IBasePopup {
+  type: PopupType.PullRequestReview
+  repository: RepositoryWithGitHubRepository
+  pullRequest: PullRequest
+  review: ValidNotificationPullRequestReview
+  numberOfComments: number
+  shouldCheckoutBranch: boolean
+  shouldChangeRepository: boolean
+}
+
+interface IUnreachableCommitsPopup extends IBasePopup {
+  type: PopupType.UnreachableCommits
+  selectedTab: UnreachableCommitsTab
+}
+
+interface IStartPullRequestPopup extends IBasePopup {
+  type: PopupType.StartPullRequest
+  allBranches: ReadonlyArray<Branch>
+  currentBranch: Branch
+  defaultBranch: Branch | null
+  externalEditorLabel?: string
+  imageDiffType: ImageDiffType
+  recentBranches: ReadonlyArray<Branch>
+  repository: Repository
+  nonLocalCommitSHA: string | null
+  showSideBySideDiff: boolean
 }
 
 export type Popup =
-  | { type: PopupType.RenameBranch; repository: Repository; branch: Branch }
```

Comment:
I assume you made all these changes so that all popup types had the common `id` property.

If you want to keep the changes to a minimum, I think you can write something like this:

```typescript
type PopupDetail = PopupType1 | PopupType2 | ...
export type Popup = { readonly id?: string } & PopupDetail
```

This is an example of using `&` to "join" types

Reply:
I think my druthers would be to have done interfaces.. but I have no real reasoning asides from it looks/feels nicer which is probably just "what I am used to" so not really valid :D ; and in this case I love not having to change as much existing stuff.

Reply:
Yeah I think deciding between types and interfaces is tough. I'd probably had done the same as you… for a new file. For an existing long file like this I'd rather keep the changes to a minimum. But that still is a personal preference! 😂